### DISCUSSION TITLE
Responsive canvas items

### DISCUSSION THREAD

**Discussion by khaledmndev** [fabricjs/fabric.js#7977] [Q&A]
Hello All,
i'm having some issues when the canvas width and height changes, i want the items to be in there same position..

i' using the latest version of fabricjs and using pdfjs library with it..

here's the code i'm using to try to fix it, but not working unfortunately:

```
function rescale_canvas_if_needed() {
  var optimal_dimensions = [$(".canvasWrapper").outerWidth(), $(".canvasWrapper").outerHeight()];
  var scaleFactorX = window.innerWidth / optimal_dimensions[0];
  var scaleFactorY = window.innerHeight / optimal_dimensions[1];
  if (scaleFactorX < scaleFactorY && scaleFactorX < 1) {
    canvas.setWidth(optimal_dimensions[0] * scaleFactorX);
    canvas.setHeight(optimal_dimensions[1] * scaleFactorX);
    canvas.setZoom(scaleFactorX);
  } else if (scaleFactorX > scaleFactorY && scaleFactorY < 1) {
    canvas.setWidth(optimal_dimensions[0] * scaleFactorY);
    canvas.setHeight(optimal_dimensions[1] * scaleFactorY);
    canvas.setZoom(scaleFactorY);
  } else {
    canvas.setWidth(optimal_dimensions[0]);
    canvas.setHeight(optimal_dimensions[1]);
    canvas.setZoom(1);
  }

  canvas.calcOffset();
  canvas.renderAll();
}


function handle_resize() {
  $(".canvas-container").hide();
  rescale_canvas_if_needed();
  $(".canvas-container").show();

}
```
 

here's the live webapp i made: https://khaledmndev.github.io/testProject/peditor/web/  (wait for the pdf file to load please)..

try adding an drawing on the book and click on the zoom in and out and see how the canvas items  position is changing

**Comment by c-diego** [comment]
Hello @khaledmndev, how are you?

I had the same problem in my project. When the zoom was changed, the annotations (fabric objects) were moved to another part of the canvas (in fact, they remained in the same place (e.g. 100px top, 80px left), the height and width of the canvas were changed). To solve my problem I had to change the position and size of the annotations (fabric objects).

To change them I took the current zoom and divided it by the previous zoom. (by zoom, understand that it is the scale passed in `getVieweport({scale: 1.5})`)

After that I iterated over the list that contained all the objects and changed them one by one.

```ts
const scaleFactor = currentZoom / previousZoom;

/** Atualizando a posição e tamanho das anotações caso seja necessário */
annotations.current.forEach((annotation) => {
    if (annotation instanceof IText) {
        annotation.set({
            fontSize: annotation.fontSize * scaleFactor,
        });
    }

    if (annotation instanceof Group) {
        annotation.set({
            scaleX: annotation.scaleX * scaleFactor,
            scaleY: annotation.scaleY * scaleFactor,
        });
    }

    if (annotation instanceof Rect) {
        annotation.set({
            width: annotation.width * scaleFactor,
            height: annotation.height * scaleFactor,
        });
    }

    annotation.set({
        left: annotation.left * scaleFactor,
        top: annotation.top * scaleFactor,
    });

    newFabric.add(annotation);
}
```
